### DISCUSSION TITLE
How to get all subscribed topics?

### DISCUSSION THREAD

**Discussion by ksetrin** [invertase/react-native-firebase#4819] [Q&A] [answered]
# Question

I use topics on my project and sometimes users tell me they have a push messages not related to user. 
This happens when a user logs out and logs in to another account
And finally I fixed that issue, I started to keep subscribed topics in AsyncStorage

But, what if a user clear all data (android)?
Do I stay subscribed to the push notifications?

What if unsubscribe processing will be terminated (or early), but user already logged out?
This will mean that the user is already logged out but still subscribed to topics

**The question is: How to know all subscribed topics in any time?**

I found one way to [Get all subscribed topics from firebase cloud messaging](https://stackoverflow.com/questions/37987821/get-all-subscribed-topics-from-firebase-cloud-messaging) from stackOverflow, but this method does not work for iOS (not tested for Android, will do it soon)

So, assume it works for android then how to get the same for iOS?

**Comment by JoeToeniskoetter** [comment, marked as answer]
I believe using this google apis endpoint https://iid.googleapis.com/iid/info/<TOKEN>?details=true

Should return a list of subscribed topics regardless of platform (iOS, android) for a particular FCM token. 

My suggestion would be to create some server side logic either in a cloud function or Http endpoint that hits this endpoint with your Firebase server key, 
To get a list of topics a user is subscribed to, then loops through each subscribed topic and call 

`admin.messaging().unsubscribeFromTopic(registrationTokens, topic)`

Passing in the users FCM token and topic be unsubscribed from. 

You can then make a request to this endpoint or cloud function when a user signs out of your application.

**Reply by ksetrin** [reply]
Ok, I tested right now `https://iid.googleapis.com/iid/info/${token}?details=true` and what I get

iOS
```
{
   "applicationVersion":"1.0.4",
   "application":"xx.xxx",
   "scope":"*",
   "authorizedEntity":"XXX989301XXX",
   "rel":{
      "topics":{
         "pupil-XXX-6":{
            "addDate":"2021-01-26"
         },
         "pupil-XXX-5":{
            "addDate":"2021-01-26"
         },
         "pupil-XXX-4":{
            "addDate":"2021-01-26"
         }
      }
   },
   "platform":"IOS"
}
```

Android
```
{
   "applicationVersion":"80",
   "application":"xx.xxx",
   "scope":"*",
   "authorizedEntity":"XXX989301XXX",
   "rel":{
      "topics":{
         "topic-XXX-6":{
            "addDate":"2021-01-25"
         },
         "topic-XXX-5":{
            "addDate":"2021-01-25"
         },
         "topic-XXX-4":{
            "addDate":"2021-01-25"
         }
      }
   },
   "appSigner":"XXXX16062ea3cd2c4aXXXX7876baa6f38cabXXXX",
   "platform":"ANDROID"
}
```
@JoeToeniskoetter
So, you are right, it works for both and it's enough
Thank you for answer

**Reply by ksetrin** [reply]
But one correction for 

`admin.messaging().unsubscribeFromTopic(registrationTokens, topic)`

https://rnfirebase.io/reference/messaging#unsubscribeFromTopic

`unsubscribeFromTopic`
Unsubscribe the device from a topic.

```unsubscribeFromTopic(topic: string): Promise<void>;```
Only accepts 1 parameter


P.S The correction for those who find this question in future

**Reply by JoeToeniskoetter** [reply]
@ksetrin 

I should've specified that the `admin.messaging().unsubscribeFromTopic(registrationTokens, topic)` function was coming from the server library `firebase-admin` not the react native library as this logic needs to exist on the server to keep your server key private. Do not make api requests with your server key from the app itself as this is not secure.

https://firebase.google.com/docs/cloud-messaging/manage-topics

```
// Unsubscribe the devices corresponding to the registration tokens from
// the topic.
admin.messaging().unsubscribeFromTopic(registrationTokens, topic)
```

So if you were using a server framework like express, you could create an endpoint that accepts the user's FCM token and unsubscribe them from all topics on the server side.

This code is untested..

```typescript
import express from 'express';
import admin from 'firebase-admin';
import fetch from 'isomorphic-unfetch';

app.post('/unsubscribe', async (req:Request, res:Response, next:NextFunction) => {
    const token = request.body.token;
    if(!token || token.trim() === ""){
        return res.status(400).json({message:"invalid token");
    }

    const result = await fetch(`https://iid.googleapis.com/iid/info/${token}?details=true`, {
      headers:{
        'Authorization': `Bearer ${process.env.SERVER_KEY}`,
        'content-type': 'application/json'
      }
    });

    const jsonResult = await result.json();
    //TODO add a check to make sure the rel and topics keys
    //exist on the jsonResult object

    //use your favorite higher order function to deal with the array of topics
    const topicsToUnsubscribeFrom = Object.keys(jsonResult.rel.topics);
    topicsToUnsubscribeFrom.forEach(async(topic) => {
      await admin.messaging().unsubscribeFromTopic(token, topic);
    });

    res.json({message:'ok'});
    
}

app.listen(process.env.PORT || 3000);

```

**Reply by Nashmod** [reply]
How do I get the token of the user from the react native app so I can make the call in the app to retrieve the subscribed topics?

**Reply by mikehardy** [reply]
@Nashmod users don't have tokens. A specific installation of your app on a specific device has a token. Either the messaging.getToken API or the Installations package has an API for obtaining this token. Managing that token in relation to users and devices and your business needs is project-specific. There are some good guides on token handling provided from the firebase docs / guides upstream

**Reply by Nashmod** [reply]
@mikehardy Ah i understand. I think i was wording my question wrong. I did mean token for the specific device but you did answer my question thank you! 
This is my first discussion post so I'm glad to see a quick response and having it resolved 😁😁

**Reply by rohan20** [reply]
Now that this endpoint has been deprecated, is there an alternative without having to maintain the list of topics on our own server?

**Reply by mikehardy** [reply]
There is no API to list all available topics
There is no API to list the topics that a given FCM token is subscribed to

Those are true and complete statements, and the implications are left to the project as what you need to do based on those two facts is different depending on your project needs.

It is worth mentioning that a way to "unsubscribe all" for a given token is to render the token invalid by calling https://rnfirebase.io/reference/installations#delete then https://rnfirebase.io/reference/installations#getToken to get a new token (or similarly https://rnfirebase.io/reference/messaging#deleteToken + https://rnfirebase.io/reference/messaging#getToken )

If you use FCM you should already be managing token staleness serverside such that you will see FCM send failures to the old / invalid token and remove it server side in response (https://firebase.google.com/docs/cloud-messaging/manage-tokens)

If you have project-specific requirements so know at all times what topics a given token is subscribed to then you must maintain that map in your project data somewhere. If you must know that to a user level (where a user may have multiple tokens corresponding to multiple installations) then you'll need to manage that user->token mapping as well as token->topic map
